bcdaworker/worker/alr.go

```diff
@@ -0,0 +1,215 @@
+package worker
+
+import (
+	"bufio"
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/CMSgov/bcda-app/bcda/models"
+	"github.com/CMSgov/bcda-app/bcda/models/fhir/alr"
+	"github.com/CMSgov/bcda-app/bcda/models/postgres"
+	"github.com/CMSgov/bcda-app/bcda/utils"
+	"github.com/CMSgov/bcda-app/bcdaworker/repository"
+	"github.com/CMSgov/bcda-app/conf"
+	"github.com/google/fhir/go/jsonformat"
+	"github.com/google/fhir/go/proto/google/fhir/proto/stu3/resources_go_proto"
+	"github.com/huandu/go-sqlbuilder"
+	"github.com/pborman/uuid"
+	"github.com/sirupsen/logrus"
+)
+
+type AlrWorker interface {
+	ProcessAlrJob(ctx context.Context, jobArgs models.JobAlrEnqueueArgs) error
+	UpdateJobAlrStatus(ctx context.Context, jobID uint, newStatus models.JobStatus) error
+	GetAlrJobStatus(ctx context.Context, jobID uint) (*models.JobStatus, error)
+}
+
+type alrWorker struct {
+	*postgres.AlrRepository
+	FHIR_STAGING_DIR string
+	ndjsonFilename   string
+}
+
+const (
+	sqlFlavor = sqlbuilder.PostgreSQL
+)
+
+func NewAlrWorker(db *sql.DB) *alrWorker {
+	alrR := postgres.NewAlrRepo(db)
+
+	// embed data struct that has method GetAlr
+	worker := &alrWorker{
+		AlrRepository:    alrR,
+		FHIR_STAGING_DIR: "",
+		ndjsonFilename:   "", // Filled in later
+	}
+
+	err := conf.Checkout(worker) // worker is already a reference, no & needed
+	if err != nil {
+		logrus.Fatal("Could not get data from conf for ALR.", err)
+	}
+
+	return worker
+}
+
+func (a *alrWorker) ProcessAlrJob(
+	ctx context.Context,
+	jobArgs models.JobAlrEnqueueArgs,
+) error {
+
+	// Parse the jobAlrEnqueueArgs
+	aco := jobArgs.CMSID
+	id := jobArgs.ID
+	MBIs := jobArgs.MBIs
+	lowerBound := jobArgs.LowerBound
+	upperBound := jobArgs.UpperBound
+
+	// Pull the data from ALR tables (alr & alr_meta)
+	alrModels, err := a.GetAlr(ctx, aco, MBIs, lowerBound, upperBound)
+	if err != nil {
+		logrus.Error(err)
+		return err
+	}
+
+	// Set up IO operation to dump ndjson
+	fileUUID := uuid.New()
+	f, err := os.Create(fmt.Sprintf("%s/%d/%s.ndjson", a.FHIR_STAGING_DIR,
+		id, fileUUID))
+	if err != nil {
+		logrus.Error(err)
+		return err
+	}
+	w := bufio.NewWriter(f)
+	defer utils.CloseFileAndLogError(f)
+
+	// Serialize data into JSON
+	marshaller, err := jsonformat.NewPrettyMarshaller(jsonformat.STU3)
```

Comment:
I think we should use the std marshaller. Since we are returning NDJSON files, we want every ALR record to be contained in a single line.